Module 2

simon-towler opened the pull request:
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

Commit:
Add Trello instructions to README

Commit:
Add external lib Requests to TOML file

Commit:
Get items from Trello instead of from cookie

Commit:
Handle the fetched items as JSON

Their IDs display. Still work to do to get their status and title.

Commit:
Get the items' name/title

Commit:
Use closed status (true/false) as status of items

Commit:
Add an item to Trello

TODO the page should redisplay correctly after new item is added

Commit:
Fix redisplay of To-Do App index page after new item submitted

Commit:
Get Trello key and token from the environment

Commit:
Use variables for Trello list ids

Commit:
Change the status of items to Done

Add a "complete" button to each item in the list of items.
When the button is clicked the item moves to the Done list.
The item is not displayed. (The Done list is not displayed.)
TODO It would be better to see both lists, To Do and Done.
TODO To Do items are in the wrong Trello list, Doing instead of To Do.

Commit:
Use Trello To Do list for To Dos, instead of Trello Doing list

Commit:
Show both ToDo and Done lists, move items between them and back

Use corresponding Trello boards.

Commit:
Add an Item class

KaraSDavis commented on todo_app/app.py:
```diff
@@ -1,20 +1,52 @@
+TRELLO_TOKEN = os.environ.get('TRELLO_TOKEN')
+TRELLO_LIST_ID_DOING = "609392e0aaa6e8618e341f90"
+TRELLO_LIST_ID_TODO = "609392e0aaa6e8618e341f8f"
+TRELLO_LIST_ID_DONE = "609392e0aaa6e8618e341f91"
```
It's nice to store these in the .env file as well, and add instructions for getting the correct IDs from Trello to your readme - that way a new user just has to update .env to get running rather than changing any of the .py files.

KaraSDavis commented on todo_app/app.py:
```diff
@@ -1,20 +1,52 @@
-    return render_template('index.html', items=get_items())
+    todos = requests.get(f"https://trello.com/1/lists/{TRELLO_LIST_ID_TODO}/cards?key={TRELLO_KEY}&token={TRELLO_TOKEN}")
+    dones = requests.get(f"https://trello.com/1/lists/{TRELLO_LIST_ID_DONE}/cards?key={TRELLO_KEY}&token={TRELLO_TOKEN}")
+    return render_template('index.html', todos=todos.json(), dones=dones.json())
```
Instead of passing the json directly into the template, you can loop through these two lists and create an instance of `Item` for each element, setting its name and status appropriately. Then in `index.html` you can display that status instead of using `todo.closed`.

KaraSDavis reviewed (changes requested):
Nice work! :) Everything seems to function correctly, I've just left a couple of comments with things you can neaten up a little. You should also be able to remove a few things that are no longer needed - the `SECRET_KEY` in .env and the `session_items.py`.

Once you've implemented the change to make use of the Item class I'll be happy to approve this.

Commit:
Use an Item class for list items.

Get Trello constants from the environment.
Add Trello constants to the .env file.
Add instructions for getting value of Trello constants to README.
Remove SECRET_KEY from .env file.

hugh-emerson commented on todo_app/app.py:
```diff
@@ -1,20 +1,63 @@
+    todos = requests.get(f"https://trello.com/1/lists/{TRELLO_LIST_ID_TODO}/cards?key={TRELLO_KEY}&token={TRELLO_TOKEN}")
+    dones = requests.get(f"https://trello.com/1/lists/{TRELLO_LIST_ID_DONE}/cards?key={TRELLO_KEY}&token={TRELLO_TOKEN}")
+    # loop through these two lists and create an instance of Item for each element
+    todos = itemizeList(todos.json())
+    dones = itemizeList(dones.json())
```
Note that the Module 3 exercise assumes you get a single list of all of the cards, and then asks you to sort them into separate lists inside a "ViewModel".

You're already getting your items in separate lists so the exercise's suggestion doesn't make a lot of sense.

You probably want to just ignore that part but the stretch goal after it would still give you some practice writing unit tests. E.g. adding a "recent_done_items" property to give you all of the cards that were moved to "done" today.

hugh-emerson reviewed (approved):
Looks good to me! If you re-submit it in Aptem I'd be happy to mark it "complete" there too.
